src/model_instance_state.cu

```diff
@@ -33,54 +37,81 @@ ModelInstanceState::Create(
     ModelState* model_state, TRITONBACKEND_ModelInstance* triton_model_instance)
 {
   std::string instance_name = get_model_instance_name(*triton_model_instance);
-  TRITONSERVER_InstanceGroupKind instance_kind =
-      get_instance_kind(*triton_model_instance);
   int32_t instance_id = get_device_id(*triton_model_instance);
 
   return std::make_unique<ModelInstanceState>(
-      model_state, triton_model_instance, instance_name.c_str(), instance_kind,
-      instance_id);
+      model_state, triton_model_instance, instance_name.c_str(), instance_id);
 }
 
-raft::handle_t&
+std::optional<raft::handle_t>&
 ModelInstanceState::get_raft_handle()
 {
-  return *handle;
+  return handle;
 }
 
 void
 ModelInstanceState::predict(
     TritonTensor<const float>& data, TritonTensor<float>& preds,
     bool predict_proba)
 {
-  try {
-    ML::fil::predict(
-        *handle, fil_forest, preds.data(), data.data(), data.shape()[0],
-        predict_proba);
-  }
-  catch (raft::cuda_error& err) {
+  if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+    try {
+      ML::fil::predict(
+          handle.value(), fil_forest, preds.data(), data.data(),
+          data.shape()[0], predict_proba);
+    }
+    catch (raft::cuda_error& err) {
+      throw TritonException(
+          TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL, err.what());
+    }
+    catch (std::bad_optional_access& err) {
+      throw TritonException(
+          TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL,
+          "No RAFT handle created in GPU instance");
+    }
+  } else if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_CPU) {
+    gtil_predict(*this, data, preds, predict_proba);
+  } else {
     throw TritonException(
-        TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL, err.what());
+        TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INVALID_ARG,
+        "Instance kind must be set to either GPU or CPU");
   }
 }
 
 ModelInstanceState::ModelInstanceState(
     ModelState* model_state, TRITONBACKEND_ModelInstance* triton_model_instance,
-    const char* name, const TRITONSERVER_InstanceGroupKind kind,
-    const int32_t device_id)
+    const char* name, const int32_t device_id)
     : BackendModelInstance(model_state, triton_model_instance),
-      model_state_(model_state), handle(std::make_unique<raft::handle_t>())
+      model_state_(model_state), handle([&]() -> std::optional<raft::handle_t> {
+        if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+          return std::make_optional<raft::handle_t>();
+        } else {
+          return std::nullopt;
+        }
+      }())
 {
   model_state_->LoadModel(ArtifactFilename(), Kind(), DeviceId());
-  ML::fil::from_treelite(
-      *handle, &fil_forest, model_state_->treelite_handle,
-      &(model_state_->tl_params));
+
+  if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+    LOG_MESSAGE(TRITONSERVER_LOG_INFO, "Using GPU for inference");
```

Comment:
Let's make this and the next log message slightly more specific to clarify that we're using GPU for inference with a particular model or at least that this applies to only the FIL backend